libraries/cms/application/cms.php

```diff
@@ -114,103 +114,15 @@ public function __construct(JInput $input = null, Registry $config = null, JAppl
 		}
 
 		// Enable sessions by default.
-		if (is_null($this->config->get('session')))
+		if (is_null($this->getContainer()->get('config')->get('session')))
```

Comment:
You should never grab the container off the application like that! This completely beats the purpose of having a container. When you create your component (i.e. your component's container) you pass it the main container. The component MUST NOT go through the application to get the container or the application configuration. This is called tight coupling and it's been the bane of our existence for ten years.

Instead, the application must be in the container for the same reason. The component, for example, only knows about the container. Therefore the component must be able to do $container->app (if you're using magic __get) or $container->get('app') (if you're not using magic __get and note that it's NOT a a bespoke method call).

Moreover, it should be possible for me to tell the application to use a different database connection or session object than the one we normally use in Joomla!. How do we do that? By setting up a container with a different db and session object and passing that to the application constructor. Or by creating the application through the container in which case I am also passing the container (self) to the application constructor.

In fact, what you're trying to do is all backwards because the application tries to act as a DIC to itself but does that in an inflexible and clumsy way, making it MUCH worse than using JFactory. At least with JFactory we could try to override its objects using reflection in one place. Now we have to do that with every app instance which is far worse. Completely ack basswards ;) I'll let @nibra explain this in CS terms or @mbabker in practical terms, but we seriously can't do what you're proposing here. It's really bad architecture.